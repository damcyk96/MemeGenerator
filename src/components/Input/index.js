// @flow

import Animated from 'react-native-reanimated'
import React, { type ElementRef } from 'react'
import colorPackage from 'color'
import { StyleSheet, TextInput, View } from 'react-native'

import { black, darkRed, fontColor, primary, white } from 'src/styles/colors'
import { timingAnimationConfig } from 'src/helpers/animationConfigs'

const { timing, interpolate, Value } = Animated

type Props = {
  error?: string,
  label: string,
}

type State = {
  focused: boolean,
}

export class Input extends React.PureComponent<Props, State> {
  state = {
    focused: false,
  }

  _focused = new Value(0)
  _error = new Value(0)
  // $FlowFixMe
  _textInputRef: ElementRef<TextInput> = React.createRef()

  componentDidUpdate = () => {
    const newErrorValue = Number(!!this.props.error)
    timing(this._error, timingAnimationConfig(150, newErrorValue)).start()
  }

  _handleFocus = () => {
    timing(this._focused, timingAnimationConfig(150, 1)).start()
    this.setState({ focused: true })
  }

  _handleBlur = () => {
    if (!this._textInputRef.current._lastNativeText) {
      timing(this._focused, timingAnimationConfig(150, 0)).start()
    }
    this.setState({ focused: false })
  }

  focus = () => this._textInputRef.current.focus()

  render() {
    const { label, error, ...restProps } = this.props
    const color: string = error
      ? darkRed
      : this.state.focused
      ? primary
      : colorPackage(fontColor)
          .darken(0.2)
          .rgb()
          .string()

    return (
      <View style={[styles.container, { borderColor: color }]}>
        <Animated.Text
          pointerEvents="none"
          style={[
            styles.label,
            {
              color,
              fontSize: interpolate(this._focused, {
                inputRange: [0, 1],
                outputRange: [16, 17],
              }),
              transform: [
                {
                  translateY: interpolate(this._focused, {
                    inputRange: [0, 1],
                    outputRange: [13, -12],
                  }),
                  translateX: interpolate(this._focused, {
                    inputRange: [0, 1],
                    outputRange: [14, 7],
                  }),
                },
              ],
            },
          ]}
        >
          {label}
        </Animated.Text>

        <Animated.Text
          pointerEvents="none"
          style={[
            styles.error,
            {
              color,
              opacity: interpolate(this._error, {
                inputRange: [0, 1],
                outputRange: [0, 1],
              }),
              transform: [
                {
                  translateY: interpolate(this._error, {
                    inputRange: [0, 1],
                    outputRange: [40, 50],
                  }),
                  translateX: interpolate(this._error, {
                    inputRange: [0, 1],
                    outputRange: [10, 7],
                  }),
                },
              ],
            },
          ]}
        >
          {error}
        </Animated.Text>

        <TextInput
          onBlur={this._handleBlur}
          onFocus={this._handleFocus}
          ref={this._textInputRef}
          style={styles.input}
          {...restProps}
        />
      </View>
    )
  }
}

export default Input

const styles = StyleSheet.create({
  container: {
    borderColor: black,
    borderRadius: 10,
    borderWidth: StyleSheet.hairlineWidth * 2,
    marginBottom: 20,
    marginTop: 10,
    width: '95%',
  },
  error: {
    position: 'absolute',
  },
  input: {
    height: 48,
    paddingHorizontal: 12,
    width: '100%',
    zIndex: 1,
  },
  label: {
    backgroundColor: white,
    position: 'absolute',
  },
})
