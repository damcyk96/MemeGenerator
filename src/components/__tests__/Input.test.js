import React from 'react'
import renderer from 'react-test-renderer'

import Input from '../Input'

describe('<Input />', () => {
  const defaultProps = {
    label: 'Input label',
  }

  const wrapper = renderer.create(<Input {...defaultProps} />)

  test('render', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
