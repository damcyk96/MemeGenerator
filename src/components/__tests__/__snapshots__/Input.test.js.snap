// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> render 1`] = `
<View
  style={
    Array [
      Object {
        "borderColor": "#000000",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginBottom": 20,
        "marginTop": 10,
        "width": "95%",
      },
      Object {
        "borderColor": "rgb(82, 82, 82)",
      },
    ]
  }
>
  <Text
    collapsable={false}
    pointerEvents="none"
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "color": "rgb(82, 82, 82)",
        "position": "absolute",
      }
    }
  >
    Input label
  </Text>
  <Text
    collapsable={false}
    pointerEvents="none"
    style={
      Object {
        "color": "rgb(82, 82, 82)",
        "position": "absolute",
      }
    }
  />
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Object {
        "height": 48,
        "paddingHorizontal": 12,
        "width": "100%",
        "zIndex": 1,
      }
    }
    underlineColorAndroid="transparent"
  />
</View>
`;
